File: zad2 - Xmodem/src/main/java/org/example/XModemSender.java
package org.example;

import java.io.*;

public class XModemSender {

    // STX, SOH, EOT and ACK characters
    private static final byte STX = 0x02;
    private static final byte SOH = 0x01;
    private static final byte EOT = 0x04;
    private static final byte ACK = 0x06;

    // XModem packet size
    private static final int PACKET_SIZE = 128;

    // Wait time for ACK response in milliseconds
    private static final int ACK_TIMEOUT = 1000;

    // XModem packet number
    private byte packetNumber = 1;

    // Input and output streams
    private InputStream inputStream;
    private OutputStream outputStream;

    public XModemSender(InputStream inputStream, OutputStream outputStream) {
        this.inputStream = inputStream;
        this.outputStream = outputStream;
    }

    public void sendFile(String fileName) throws IOException {
        File file = new File(fileName);
        FileInputStream fileInputStream = new FileInputStream(file);

        // Send file name
        sendPacket(file.getName().getBytes());

        // Send file size
        sendPacket(String.valueOf(file.length()).getBytes());

        // Send file contents
        byte[] buffer = new byte[PACKET_SIZE];
        int bytesRead = 0;

        while ((bytesRead = fileInputStream.read(buffer)) > 0) {
            sendPacket(buffer, bytesRead);
        }

        // Send end of transmission
        sendPacket(new byte[] { EOT });

        fileInputStream.close();
    }

    private void sendPacket(byte[] data) throws IOException {
        sendPacket(data, data.length);
    }

    private void sendPacket(byte[] data, int length) throws IOException {
        byte[] packet = new byte[PACKET_SIZE + 4];

        if (length == PACKET_SIZE) {
            packet[0] = STX;
        } else {
            packet[0] = SOH;
        }

        packet[1] = packetNumber;
        packet[2] = (byte) (~packetNumber);

        System.arraycopy(data, 0, packet, 3, length);

        byte checksum = 0;
        for (int i = 3; i < PACKET_SIZE + 3; i++) {
            checksum += packet[i];
        }
        packet[PACKET_SIZE + 3] = checksum;

        boolean ackReceived = false;
        while (!ackReceived) {
            outputStream.write(packet);
            outputStream.flush();

            byte[] response = new byte[1];
            long startTime = System.currentTimeMillis();

            while (System.currentTimeMillis() - startTime < ACK_TIMEOUT) {
                if (inputStream.available() > 0) {
                    inputStream.read(response);
                    if (response[0] == ACK) {
                        ackReceived = true;
                        packetNumber++;
                        break;
                    }
                }
            }

            if (!ackReceived) {
                // Resend packet
            }
        }
    }
}

File: zad2 - Xmodem/src/XModemSender  .cpp
#include <iostream>
#include <chrono>
#include <thread>
#include "XModemSender.h"
#include "variables.h"

// Input and output streams
XModemSender::XModemSender(istream &inputStream, ostream &outputStream) : inputStream(inputStream), outputStream(outputStream) {}

void XModemSender::sendFile(const string &fileName) {
    // Open file for reading
    ifstream fileInputStream(fileName, ios::binary);
    if (!fileInputStream) {
        throw runtime_error("Unable to open file: " + fileName);
    }

    // Wait for NAK from receiver
    bool nakReceived = false;
    auto startTime = chrono::system_clock::now();
    while (chrono::duration_cast<chrono::milliseconds>(chrono::system_clock::now() - startTime).count() < 60000) {
        if (inputStream.rdbuf()->in_avail() > 0) {
            int b = inputStream.get();
            if (b == NAK) {
                nakReceived = true;
                break;
            }
        }
        this_thread::sleep_for(chrono::milliseconds(10));
    }
    if (!nakReceived) {
        throw runtime_error("Timeout waiting for NAK response from receiver");
    }

    // Send file contents
    int blockNumber = 1;
    char packetData[PACKET_SIZE];
    int bytesRead;
    while ((bytesRead = fileInputStream.readsome(packetData, PACKET_SIZE)) > 0) {
        // Send packet
        sendPacket(packetData, bytesRead, blockNumber);

        // Increment block number
        blockNumber++;

        // Wait for ACK or NAK response
        bool ackReceived = false;
        bool nakReceivedAgain = false;
        startTime = chrono::system_clock::now();
        while (chrono::duration_cast<chrono::milliseconds>(chrono::system_clock::now() - startTime).count() <
               RESPONSE_TIMEOUT) {
            if (inputStream.rdbuf()->in_avail() > 0) {
                int b = inputStream.get();
                if (b == ACK) {
                    ackReceived = true;
                    break;
                } else if (b == NAK) {
                    nakReceivedAgain = true;
                    break;
                }
            }
        }
        if (!ackReceived) {
            if (nakReceivedAgain) {
                // Resend packet
                blockNumber--;
            } else {
                throw runtime_error("Timeout waiting for ACK/NAK response from receiver");
            }
        }
    }

    // Send EOT
    sendEOT();

    // Wait for ACK response
    bool eotAckReceived = false;
    startTime = chrono::system_clock::now();
    while (chrono::duration_cast<chrono::milliseconds>(chrono::system_clock::now() - startTime).count() <
           RESPONSE_TIMEOUT) {
        if (inputStream.rdbuf()->in_avail() > 0) {
            int b = inputStream.get();
            if (b == ACK) {
                eotAckReceived = true;
                break;
            }
        }
    }
    if (!eotAckReceived) {
        throw runtime_error("Timeout waiting for ACK response from receiver after EOT");
    }

    fileInputStream.close();
}

void XModemSender::sendPacket(char data[], int length, int blockNumber) {
    // Construct packet
    char packet[PACKET_SIZE + 3];
    packet[0] = SOH; // Start of Header
    packet[1] = (char) blockNumber;
    packet[2] = (char) ~blockNumber;
    memcpy(packet + 3, data, length);

    // Calculate checksum
    char checksum = 0;
    for (int i = 0; i < PACKET_SIZE; i++) {
        checksum += packet[i + 3];
    }
    packet[PACKET_SIZE + 3 - 1] = checksum;

    // Send packet
    outputStream.write(packet, PACKET_SIZE + 3);
    outputStream.flush();
}

void XModemSender::sendEOT() {
    outputStream.write(&EOT, 1);
    outputStream.flush();
}



